dataloader/preprocessor.hpp

```diff
@@ -48,12 +48,117 @@ class PreProcessor
+   * arma::mat type for training YOLO model. Each column in target matrix has
+   * the size : gridWidth * gridHeight * (5 * numBoxes + classes).
+   *
+   * @param annotations Field object created using model's dataloader containing
+   *                    annotation for images.
+   * @param output Output matrix where output will be stored.
+   * @param imageWidth Width of image used for training YOLO model.
+   * @param imageHeight Height of image used for training YOLO model.
+   * @param gridWidth Width of output feature map of YOLO model.
+   * @param gridHeight Height of output feature map of YOLO model.
+   * @param numBoxes Number of bounding boxes per grid.
+   * @param numClasses Number of classes in training set.
+   *
+   * Note : This function must be called manually before model is used.
+   */
+  template<typename eT>
+  static void YOLOPreProcessor(const DatasetY& annotations,
+                               arma::Mat<eT>& output,
+                               const size_t imageWidth = 224,
+                               const size_t imageHeight = 224,
+                               const size_t gridWidth = 7,
+                               const size_t gridHeight = 7,
+                               const size_t numBoxes = 2,
+                               const size_t numClasses = 20)
+  {
+    mlpack::Log::Assert(typeid(annotations) == typeid(arma::field<arma::vec>),
+        "Use Field type to represent annotations.");
+    size_t batchSize = annotations.n_cols;
+    size_t numPredictions = 5 * numBoxes + numClasses;
+    double cellSizeHeight = (double) 1.0 / gridHeight;
+    double cellSizeWidth = (double) 1.0 / gridWidth;
+
+    // Set size of output and use cubes convenience.
+    output.set_size(gridWidth * gridHeight * numPredictions, batchSize);
+    output.zeros();
+
+    // Use offset to create a cube for a particular column / batch.
+    size_t offset = 0;
+    for (size_t boxIdx = 0; boxIdx < batchSize; boxIdx++)
+    {
+      arma::cube outputTemp(const_cast<arma::Mat<eT> &>(output).memptr() +
+        offset, gridHeight, gridWidth, numPredictions, false, false);
+      offset += gridWidth * gridHeight * numPredictions;
+
+      // Get the bounding box and labels corresponding to current image.
+      arma::mat labels(1, annotations(0, boxIdx).n_elem / 5);
+      arma::mat boundingBoxes(4, annotations(0, boxIdx).n_elem / 5);
+      for (size_t i = 0; i < boundingBoxes.n_cols; i++)
+      {
+        labels.col(i)(0) = annotations(0, boxIdx)(i * 5);
+        boundingBoxes.col(i) = annotations(0, boxIdx)(arma::span(i * 5 + 1,
+            (i + 1) * 5 - 1));
+      }
+
+
+      // Normalize the coordinates.
+      boundingBoxes.row(0) /= imageWidth;
+      boundingBoxes.row(2) /= imageWidth;
+      boundingBoxes.row(1) /= imageHeight;
+      boundingBoxes.row(3) /= imageHeight;
+
+      // Get width and height as well as centres for the bounding box.
+      arma::mat widthAndHeight(2, boundingBoxes.n_cols);
```

Comment:
I have seen at least one dataset that alreay used the centered bbox format but wasn't normalized, do you think we should introduce a parameter to skip the step?

Reply:
Ahh makes sense. I'll add that as well.